File: gfg/logic_builind/1.cpp
#include<bits/stdc++.h>
using namespace std;

void IsEven(int n){
    if (n&1 !=0){
        cout<<"Odd";
    }else{
        cout<<"Even";
    }
    cout<<endl;
}
void printTable(int n){
    for(int i = 1;i<=10;i++){
        cout<<n<<" * "<<i<<" = "<<n*i<<endl;
    }
}
long long findSum(int n){
    long long sum =0;
    for(int i =1;i<=n;i++){
        sum+=i;
    }
    return sum;
}
long long sumformula(int n){
    return n*1LL*(n+1)/2;
}
long long findSquareSum(int n){
    long long sum = 0;
    for(int i=1;i<=n;i++){
        sum +=i*i;
    }
    return sum;
}
long long findSquareFormula(int n){
    return ((n*1LL*(n+1)/2)*((2*n)+1))/3;
}
template<typename T>
void swapo(T &a, T &b){
    T temp = b;
    b = a;
    a = temp;
}
int closestNumber(int n, int m){
    int closest = 0;
    int minDistance = INT_MAX;
    for (int i=n-abs(m); i<= n+abs(m);i++){
        if(i%m ==0){
            // closest=max(closest, abs(i));
            if(abs(i-n)<= minDistance){
                if(abs(i)> abs(closest)) closest = i;
                minDistance = abs(i-n);
            }
        }
    }

    return closest;
}


int main(){
    // cout<<"HI";
    int n;
    cin>>n;
    IsEven(n);
    printTable(n);
    cout<<findSum(n)<<endl;
    cout<<sumformula(n)<<endl;
    cout<<findSquareSum(n)<<endl;
    cout<<findSquareFormula(n)<<endl;
    string a = "Aditya";
    string b = "Kumar";
    cout<<"Before swap "<<a<<" and "<<b<<endl;
    swapo(a, b);
    cout<<"After swap "<<a<<" and "<<b<<endl;

    int no , mo;
    cin>>no>>mo;
    cout<<closestNumber(no,mo)<<endl;
}

File: gfg/logic_builind/1.rs
use std::io;
// use std::intrinsics;
use std::mem;

fn main(){
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Faild to read line");
    let n:i32 = input.trim().parse().unwrap();

    IsEven(n);
    printTable(n);
    println!("The sum of n is {}", findSum(n));
    println!("The sum formula is {}", findsumformula(n));
    println!("The sum of Square of n is {}", findSquareSum(n));
    println!("The sum of Square of n formula {}", findSquareFormula(n));
    let mut a = 3;
    let mut b = 5;
    println!("a: {} and b: {}", a, b);
    swapo(&mut a, &mut b);
    println!("a: {} and b: {}", a, b);

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let mut iter = input.split_whitespace();

    let mut iter = input.split_whitespace();
    let n: i32 = iter.next().unwrap().parse().unwrap();
    let m: i32 = iter.next().unwrap().parse().unwrap();

    println!("The closest Number to {} is {} which is divisble by {}", n, closestNumber(n, m), m);

}


fn closestNumber(n:i32, m:i32) -> i32 {
    let mut closest:i32 = 0;
    let mut minDistance:i32 = i32::MAX;
    for i in (n - m.abs())..=(n + m.abs()) {
        if i%m == 0 {
            if (i-n).abs() <= minDistance {
                if i.abs() > closest.abs() {
                    closest = i;
                }
                minDistance = (i-n).abs();
            }
        }
    }
    closest
}

fn swapo<T> (a:&mut T, b:&mut T){
    mem::swap(a,b);
}

// fn swapo<'a, T> (a:&'a mut T, b:&'a mut T){
//     let temp = *a;
//     *a = *b;
//     *b = temp;
// }

// fn swapo<T>(x:&mut T, y:&mut T){
//     unsafe {
//         intrinsics::typed_swap_nonoverlapping(x, y);
//     }
// }


fn findSquareFormula(n:i32) -> i64{
    let n = n as i64;
    ((n*(n+1)/2)*((2*n) + 1))/3
}

fn findSquareSum(n:i32)-> i64{
    let mut sum:i64 =0;
    for i in 1..=n{
        sum += i as i64 * i as i64;
    }
    sum
}


fn findSum(n:i32)->i32 {
    let mut sum:i32 = 0;
    for i in 1..=n{
        sum +=i;
    }

    sum
}

fn findsumformula(n:i32)-> i64{
    let n = n as i64;
    n*(n+1)/2
}


fn printTable(n:i32){
    for i in 1..11 {
        println!("{} * {} = {}", n, i, n*i);
    }
}

fn IsEven(n:i32){
    if n & 1 == 0 {
        println!("Even");
    }else{
        println!("Odd");
    }
}